Output a sub-region specified by the user

sonVishal opened the pull request:
Using the `OutputRegionBounds` parameter, under the `&Output` heading, in the `parameter.par` file the user can define the region for which the output is to be written.

This region is provided in the following format
`OutputRegionBounds = xMin xMax yMin yMax zMin zMax`

rettenbs commented on src/Reader/readpar.f90:
```diff
@@ -3301,6 +3301,25 @@ SUBROUTINE readpar_output(EQN,DISC,IO,CalledFromStructCode)
+       END IF
+      IO%OutputRegionBounds(1:6) = OutputRegionBounds(1:6)
+      IF (OutputRegionBounds(2)-OutputRegionBounds(1) <= 0.0) THEN
+        logError(*) 'Please make sure the x bounds are correct'
```
This might be a problem. First, OutputRegionBounds should have a default value (probably 0, 0, 0, 0, 0, 0). Second, the default value should correspond to "select all cells".

rettenbs replied:
This might be a problem. First, OutputRegionBounds should have a default value (probably 0, 0, 0, 0, 0, 0). Second, the default value should correspond to "select all cells".

sonVishal replied:
I will implement the default case.

rettenbs commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -71,6 +71,60 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
+
+
+	// Extract elements based on the region specified
+	for (size_t i = 0; i < numTotalElems; i++) {
```
There should be a "short cut" for the default case, when all cells are selected.

sonVishal replied:
In case of "select all cells" I think it would be better to retain the old behaviour since I have changed the constructor for meshRefiner. With an if-statement for the loop and function overloading for the constructor this can be achieved.

rettenbs commented:
Hi,

I added two comments to your commit. We need to support the default case 
(selecting all cells) and we need a shortcut for this case. We do not 
need to iterate over all cells in this case.

Otherwise it looks fine.

Best regards,
Sebastian

On 10/04/2016 05:06 PM, Vishal Sontakke wrote:
> Using the `OutputRegionBounds` parameter, under the `&Output` heading, in the `parameter.par` file the user can define the region for which the output is to be written.
>
> This region is provided in the following format
> `OutputRegionBounds = xMin xMax yMin yMax zMin zMax`
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/SeisSol/SeisSol/pull/25
>
> -- Commit Summary --
>
>   * Make changes as per diff
>   * Remove extra bracket
>
> -- File Changes --
>
>     M src/Geometry/refinement/MeshRefiner.h (36)
>     M src/Numerical_aux/typesdef.f90 (2)
>     M src/Reader/readpar.f90 (23)
>     M src/ResultWriter/WaveFieldWriter.cpp (70)
>     M src/ResultWriter/WaveFieldWriter.h (25)
>     M src/ResultWriter/WaveFieldWriterC.cpp (5)
>     M src/ResultWriter/WaveFieldWriterExecutor.h (29)
>     M src/ResultWriter/inioutput_seissol.f90 (3)
>     M src/Solver/Interoperability.cpp (9)
>     M src/Solver/Interoperability.h (3)
>     M src/Solver/f_ftoc_bind_interoperability.f90 (3)
>
> -- Patch Links --
>
> https://github.com/SeisSol/SeisSol/pull/25.patch
> https://github.com/SeisSol/SeisSol/pull/25.diff
>

-- 
Sebastian Rettenberger, M.Sc.

Technical University of Munich
Department of Informatics
Chair of Scientific Computing

Boltzmannstrasse 3
85748 Garching, Germany

Tel. + 49 89 35831 7814
Fax. +49 89 289 18607

rettenbs@in.tum.de
http://www5.in.tum.de

sonVishal commented:
Hi,

I have added the default case where the entire region is selected.
This is supported by setting all the 6 variables to 0 in the parameter file.
`OutputRegionBounds = 0 0 0 0 0 0`

The commit named `Add a default case` has an error and I have fixed that in the next commit `Fix an error`.

Let me know if anything else is to be fixed.

Regards,
Vishal

rettenbs commented:
Hi,

just one small issue left: OutputRegionBounds has no default value 
assigned (as done for other variables):
https://github.com/sonVishal/SeisSol/blob/2a63f8f416c2226c5b5fbfc9ec3cd418b4efcee8/src/Reader/readpar.f90#L3208

Thus, if the value is not specified in the parameter file, there is no 
guarantee that it is actually "0 0 0 0 0 0".

Best regards,
Sebastian

On 10/05/2016 03:46 PM, Vishal Sontakke wrote:
> Hi,
>
> I have added the default case where the entire region is selected.
> This is supported by setting all the 6 variables to 0 in the parameter file.
> `OutputRegionBounds = 0 0 0 0 0 0`
>
> The commit named `Add a default case` has an error and I have fixed that in the next commit `Fix an error`.
>
> Let me know if anything else is to be fixed.
>
> Regards,
> Vishal
>

-- 
Sebastian Rettenberger, M.Sc.

Technical University of Munich
Department of Informatics
Chair of Scientific Computing

Boltzmannstrasse 3
85748 Garching, Germany

Tel. + 49 89 35831 7814
Fax. +49 89 289 18607

rettenbs@in.tum.de
http://www5.in.tum.de

sonVishal commented:
Hi,

It is now done.

Regards,
Vishal.
